[JENKINS-52161] Improve creation-date value determination

Wadeck opened the pull request:
- In case of migration we put `null` instead of now and so changed the display in Legacy token monitoring page.
- Improve also the difference in days for the Xxx day(s) ago to take calendar day in difference and not the total duration (<24h = 0 days, which seems wrong in UI).
- It's only a partial improvement for the ticket, 2 other parts are still not done yet. This is the most "urgent" feature to avoid loosing the information that "we do not know" the creation date of legacy token during migration from XML.

See [JENKINS-52161](https://issues.jenkins-ci.org/browse/JENKINS-52161).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* (internal) API token: Improve the creation date value by putting `null` during migration of legacy token from XML
* API token: Display better the difference in terms of days between the creation (or last use) date and now, taking only the number of calendar days into account.
* API token: Legacy token monitor page now differentiates `Unknown` creation date from today.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [N/A] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jsoref

Wadeck committed:
[JENKINS-52161] Improve creation-date value determination
- in case of migration we put null instead of now
- improve also the difference in days for the Xxx day(s) ago to take calendar day in difference and not the total duration (<24h = 0 days, which seems wrong in UI)

jsoref commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.jelly:
```diff
@@ -58,20 +58,26 @@ THE SOFTWARE.
+                                <j:set var="lastUseDateValue" value="${%NoLastUseValue}" />
                                 <j:if test="${legacyStats.lastUseDate != null}">
                                     <i:formatDate var="lastUseDateFormat" value="${legacyStats.lastUseDate}" type="both" dateStyle="medium" timeStyle="medium" />
+                                    <!--TODO convert to "Today", "Tomorrow", "4 days ago", "2 weeks ago", etc-->
```
Pet peeve: `etc.`

jsoref commented on core/src/main/java/hudson/Util.java:
```diff
@@ -23,6 +23,7 @@
  */
 package hudson;
 
+import com.sun.istack.internal.NotNull;
```
Is this the right import?

fcojfernandez replied:
I think it is not used, indeed.

jsoref commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1662,7 +1666,27 @@ public static int permissionsToMode(Set<PosixFilePermission> permissions) {
+     */
+    @Restricted(NoExternalUse.class)
+    public static int differenceInCalendarDay(@Nonnull Date a, @Nonnull Date b){
+        LocalDate aLocal = a.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
```
I have half a preference for considering the user's time zone instead of the system's. (This is achievable by passing both dates down through to the jelly and letting it do this math instead.)

Wadeck replied:
IIRC HTTP Headers do not contain any information about timezone. You mean using JavaScript instead of server-side rendering ?

jvz replied:
Except for time zones that have managed to skip a day recently (which hasn't happened for decades or possibly centuries IIRC), this logic would work for any time zone.

jsoref replied:
Yes, I meant JavaScript instead of server-side.

The issue isn't that the logic would work for any time zone, it's that the creature who wants to do the counting has a different perspective.

Date A is created at 1:03am New Zealand time.
https://www.timeanddate.com/worldclock/converter.html?iso=20171230T120300&p1=22&p2=195
Date B is created at 11:55pm New Zealand time on the same calendar date in New Zealand.
https://www.timeanddate.com/worldclock/converter.html?iso=20171231T105500&p1=22&p2=195

User C in New Zealand sees these two events as 0 days apart.

User D in Paris looks at these dates and sees them as 1 day apart.

jvz replied:
Oh, I see. Yeah, that would be an actual issue.

dwnusbaum replied:
To me it seems like we's want a more general fix for this, maybe [JENKINS-19887](https://issues.jenkins-ci.org/browse/JENKINS-19887) or the linked tickets there rather than introducing complexity here as a one-off case.

jsoref replied:
Yeah, I'm fine with it not being addressed here. I just want people to be aware. Hence "my preference" as opposed to any stronger request.

The reason to mention it here is that it impacts the "api" perspective. If the values are both exposed to the client, then the client side can be upgraded to get this w/o needing to rebuild the "api".

OTOH, if jenkins treats the strings as the "api" and doesn't see them as "frozen", then evolving it later and doing it server-side is fine. -- I haven't built up a complete sense of how jenkins manages its "apis". (I just know that it has things that are apis.)

Wadeck replied:
@jsoref FYI about API:

As the code that is modified here is `@Restricted` (actually an inner class of a restricted class) we can consider it as an internal API (i.e. private) and so we could break it without any other consideration.

If there are public methods in a non-`@Restricted`, they could be used by other plugins / components, that's the developer API we need to care about in terms of binary compatibility.

jsoref replied:
@Wadeck: k. that covers the java side, but what are the rules for the http side? (Maybe the answer is there aren't any constraints, which is probably a good thing here)

Wadeck replied:
@jsoref yeah in general I would say there is no constraint except that you need to keep consistency, security and compatibility in mind.

jsoref reviewed (commented):
Thanks, this seems like a good improvement over what's happening today.

Wadeck committed:
Address Josh's comments

fcojfernandez reviewed (approved):
🐝 
LGTM, just a minor comment

daniel-beck commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -371,6 +371,22 @@ private void init() {
             return result;
         }
+    
+        public static @Nonnull HashedToken buildNewFromLegacy(@Nonnull HashValue value, boolean migrationFromExistingLegacy) {
```
`@since` or `@Restricted`

Wadeck replied:
as the parent class (ApiTokenStore) is `@Restricted`, this method is also `@Restricted` by implication

dwnusbaum commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1662,7 +1665,27 @@ public static int permissionsToMode(Set<PosixFilePermission> permissions) {
+    public static int differenceInCalendarDay(@Nonnull Date a, @Nonnull Date b){
+        LocalDate aLocal = a.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
+        LocalDate bLocal = b.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
+        return Period.between(aLocal, bLocal).getDays();
```
I think `Period#getDays` just returns the number of days less than a month, not the total (i.e. differenceInCalendarDay(2017-01-01, 2018-01-01) will return 0, but I think you want it to return 365). You should use `ChronoUnit.DAYS.between(aLocal, bLocal)` to get the total number of days.

jvz replied:
Nice catch! It's due to things like this that I tend to use `Instant` more often than the human versions.

varyvol commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1662,7 +1665,27 @@ public static int permissionsToMode(Set<PosixFilePermission> permissions) {
+     * even if there are only 3 hours between. As well as "10am" to "2pm" both on the same day, returns 0.
+     */
+    @Restricted(NoExternalUse.class)
+    public static int differenceInCalendarDay(@Nonnull Date a, @Nonnull Date b){
```
Days instead of day for all these methods?

jvz replied:
How about `daysBetween()`?

jvz commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1662,7 +1665,27 @@ public static int permissionsToMode(Set<PosixFilePermission> permissions) {
+     * @see #differenceInCalendarDay(Date, Date)
+     */
+    @Restricted(NoExternalUse.class)
+    public static int numOfCalendarDayToNow(@Nonnull Date date){
```
Could name this something like `daysSince()` or `daysElapsedSince()`. Also, based on usage, you could also return 0 for future dates to simplify a bit.

jvz commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -371,6 +371,22 @@ private void init() {
             return result;
         }
+    
+        public static @Nonnull HashedToken buildNewFromLegacy(@Nonnull HashValue value, boolean migrationFromExistingLegacy) {
```
I already packed it away, so I'm not sure which item it is, but Effective Java suggests using enums instead of booleans for situations like this. The context of what exactly that flag does is lost in a language like Java (i.e., no passing parameters by name).

Wadeck replied:
In this case the method is `@Restricted` and so can be considered "private" for us, i.e. the context is never lost as we are using the source directly.

jvz commented on core/src/test/java/hudson/UtilTest.java:
```diff
@@ -781,4 +782,20 @@ public void testPermissionsToMode() throws Exception {
+        Date day2_01am = new Date(2018, 4, 7, 1, 0);
+        Date day2_11pm = new Date(2018, 4, 7, 1, 0);
+        Date day3_08am = new Date(2018, 4, 8, 8, 0);
+        assertEquals(0, Util.differenceInCalendarDay(day1_10am, day1_11pm55));
```
It might be simpler to use a parameterized test for this, but it'd have to be in a separate class thanks to junit 4.

jvz reviewed (changes requested):
Added a couple notes about naming and boolean usage, but otherwise looks good.

Wadeck committed:
Address other comments
- also add translation for another creationDate
- and manage correctly the null creation date instead of 0

jvz reviewed (approved):
LGTM

jsoref commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.jelly:
```diff
@@ -138,14 +144,14 @@ THE SOFTWARE.
                     </j:choose>
                 </table>
-    
+                
```
I'm not sure of the normal style. I personally am not a fan of trailing whitespace (and apparently the code wasn't using it before).

Wadeck replied:
Yeah I just tried to make it consistent

jsoref commented on core/src/test/java/hudson/UtilTest.java:
```diff
@@ -781,4 +781,29 @@ public void testPermissionsToMode() throws Exception {
 
+    @Test
+    public void testDifferenceDays() {
+        Date may_6_10am = new Date(2018 - 1900, 4, 6, 10, 0);
```
https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#constructor.summary

> Date(int year, int month, int date, int hrs, int min, int sec)
> **Deprecated.**
>As of JDK version 1.1, replaced by Calendar.set(year + 1900, month, date, hrs, min, sec) or GregorianCalendar(year + 1900, month, date, hrs, min, sec).

Wadeck replied:
In general I do not care really for test code to use deprecated constructors like that because it's not the code that is tested but only the tools we use to test the real things.

As you pointed out this code and take time to retrieve the documentation, I imagine it's important for you, so I changed it 👍

Wadeck committed:
Avoid using deprecated constructor in test

jsoref commented:
@Wadeck: thanks, that's actually *much* easier to read.

Wadeck commented:
@reviewbybees done

@jsoref do you approve this PR ? :)
